### PR TITLE
[MLX] Gemma4-31B ondevice sampling

### PR DESCRIPTION
**Pull request by kiymetakdemir**
Summary

 Lets the MLX-exported Gemma 4 31B model sample the next token on-device instead of returning logits for host-side sampling. Sampling is opt-in at export (--sample); temperature, top_k, top_p, and seed are runtime inputs, and the runner increments the seed per token.

  Changes

  - export.py — --sample wraps the model in SamplingHead so forward(tokens, input_pos, temperature, top_k, top_p, seed) → int64 token; records a use_sampling constant-method flag. Non-sample export unchanged.
  - gemma4_31b_engine.cpp — reads use_sampling from metadata; when set, consumes the int64 token directly instead of logits_to_token, feeds the scalar inputs (temperature, top_k, top_p, seed) in prefill/decode (across the min/max prefill chunking), and manages the per-token seed schedule. top_k (0 = keep all → INT64_MAX on-device) and top_p are now supported; top_p is range-checked to (0, 1]; top_k/top_p/seed are rejected on non-sample models.
  - main.cpp — --top_k / --top_p / --seed flags wired into SamplingConfig; an unset seed is randomized only for sampling models.
  - tests/test_mlx_pipeline.py — test_export_to_pte_with_sampling: tiny-model MLX export with --sample, asserting the use_sampling flag, int64 token output, and same-seed reproducibility.

### CODE REVIEW

File: examples/models/gemma4_31b/export.py [hunk 1]
```diff
@@ -306,11 +309,30 @@ def _export_cuda(
     print("Done.")
 
 
+class _MLXSampleWrapper(nn.Module):
```
Comment (metascroy) [inline examples/models/gemma4_31b/export.py]: Ideally SamplingHead from backends/mlx/llm would be directly usable. The fact that we keep defining wrappers suggest that maybe it should be refactored to work?

In terms of shape, I expect sampling to work on (B, vocab) and return (B) (tokens).

I don't see a case for (B, S, vocab).

**Comment by metascroy** [comment]
@claude review this code

File: examples/models/gemma4_31b/export.py (hunk 1 above)
Comment (kiymetakdemir) [reply examples/models/gemma4_31b/export.py]: I've made SamplingHead operate on (B, vocab) → (B) so it's model-agnostic, but it still needs a positional adapter for torch.export since its sampling args are keyword-only.

SamplingHead wraps the whole model (self.model(*args)) behind a generic, keyword-only forward(*args, temperature, seed, …), so I think each model needs a small adapter that maps its fixed forward(tokens, input_pos, temperature, top_p, seed) onto that signature.

I checked the existing heads and they avoid this by being tensor layers, lm_head (nn.Linear), FlowMatchingHead take a tensor in tensor out and are called inside the model's forward, so they ride on the model's already-positional export entry point. Would we want to change SamplingHead to a tensor head like this? The CUDA path avoids it too with a plain free function sampler.py:sample(logits, temperature) is called inside model.forward (return sample(logits, temperature)).

Comment (metascroy) [reply examples/models/gemma4_31b/export.py]: What if we did:

```
def forward(self, *args, **kwargs):
    *model_args, temperature, top_k, top_p, seed = args
    logits = self.model(*model_args, **kwargs) # [B, V]
    return torch.ops.mlx.sample(last, temperature, top_k, top_p, seed)
```

Then in Qwen code (and here), we could just do:

```
model = SamplingHead(model)
example_args = (example_tokens, example_input_pos, temp, top_k, top_p, seed)
```

With no wrapper?

Comment (kiymetakdemir) [reply examples/models/gemma4_31b/export.py]: This should work, thank you!

File: examples/models/gemma4_31b/gemma4_31b_engine.cpp [hunk 2]
```diff
@@ -326,15 +355,24 @@ class Gemma4_31BSession : public LLMSession {
...
+    if (use_sampling_) {
+      seed_ += 1;
+    }
+#endif
     return Error::Ok;
   }
 
   Result<DecodeResult> decode_one(const SamplingConfig& sampling) override {
-    if (sampling.top_p != 1.0f || sampling.top_k != 0 || sampling.seed != 0) {
+    if (sampling.top_k != 0) {
```
Comment (metascroy) [inline examples/models/gemma4_31b/gemma4_31b_engine.cpp]: I think topk is supported now?

**Review by metascroy — APPROVED** [review]

File: examples/models/gemma4_31b/gemma4_31b_engine.cpp (hunk 2 above)
Comment (kiymetakdemir) [reply examples/models/gemma4_31b/gemma4_31b_engine.cpp]: Yes, I will update this.
